serviceLibrary/src/main/java/info/mqtt/android/service/ping/AlarmPingSender.kt

```diff
@@ -2,154 +2,55 @@ package info.mqtt.android.service.ping
-import org.eclipse.paho.client.mqttv3.IMqttActionListener
-import org.eclipse.paho.client.mqttv3.MqttException
-import java.lang.Exception
-import android.annotation.SuppressLint
-import android.app.Service
-import android.content.Context
-import android.os.PowerManager
+import androidx.work.*
 import info.mqtt.android.service.MqttService
-import info.mqtt.android.service.MqttServiceConstants
-import kotlinx.coroutines.*
-import kotlin.system.measureTimeMillis
+
+import java.util.concurrent.TimeUnit
+
 
 /**
  * Default ping sender implementation on Android. It is based on AlarmManager.
  *
- * This class implements the [MqttPingSender] pinger interface
+ * This class implements the [MqttPingSender] ping interface
  * allowing applications to send ping packet to server every keep alive interval.
  *
  * @see MqttPingSender
  */
 internal class AlarmPingSender(val service: MqttService) : MqttPingSender {
+    private var continuation: Operation? = null
     private var clientComms: ClientComms? = null
-    private var alarmReceiver: BroadcastReceiver? = null
-    private var pendingIntent: PendingIntent? = null
-
-    @Volatile
-    private var hasStarted = false
+    private val workManager = WorkManager.getInstance(service)
 
     override fun init(comms: ClientComms) {
         this.clientComms = comms
-        alarmReceiver = AlarmReceiver()
     }
 
     override fun start() {
-        val action = MqttServiceConstants.PING_SENDER + clientComms!!.client.clientId
-        Timber.d("Register AlarmReceiver to MqttService$action")
-        service.registerReceiver(alarmReceiver, IntentFilter(action))
-        pendingIntent = PendingIntent.getBroadcast(service, 0, Intent(action), PendingIntent.FLAG_UPDATE_CURRENT)
-        schedule(clientComms!!.keepAlive)
-        hasStarted = true
+        val pingRepeatWorkRequest = PeriodicWorkRequest
+            .Builder(PingWorker::class.java, clientComms!!.keepAlive / 100, TimeUnit.MILLISECONDS)
+            .addTag(PING_WORKER_TAG)
+            .setConstraints(Constraints.Builder().setRequiredNetworkType(NetworkType.METERED).build())
+            .build()
+
+        continuation = workManager
+            .enqueueUniquePeriodicWork(PING_JOB, ExistingPeriodicWorkPolicy.REPLACE, pingRepeatWorkRequest)
     }
 
     override fun stop() {
         Timber.d("Unregister AlarmReceiver to MqttService ${clientComms!!.client.clientId}")
-        if (hasStarted) {
-            if (pendingIntent != null) {
-                // Cancel Alarm.
-                val alarmManager = service.getSystemService(Service.ALARM_SERVICE) as AlarmManager
-                alarmManager.cancel(pendingIntent)
-            }
-            hasStarted = false
-            try {
-                service.unregisterReceiver(alarmReceiver)
-            } catch (e: IllegalArgumentException) {
-                //Ignore unregister errors.
-            }
-        }
+        workManager.cancelAllWorkByTag(PING_WORKER_TAG)
```

Comment:
Just `cancelUniqueWork(PING_JOB)` it works

Reply:
Is this the desired change 2692bc87 ?

Reply:
Yes 👍🏽